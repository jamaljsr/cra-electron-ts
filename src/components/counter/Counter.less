.counter-body {
  padding: 20px;
  text-align: center;

  h1 {
    font-size: 5rem;
  }

  .btn-group button {
    margin: 5px;
  }
}
