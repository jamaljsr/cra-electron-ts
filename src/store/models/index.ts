import { AnyAction } from 'redux';
import { connectRouter, RouterState } from 'connected-react-router';
import { History } from 'history';
import { reducer, Reducer } from 'easy-peasy';
import counterModel, { CounterModel } from './counter';

export interface RootModel {
  router: Reducer<RouterState, AnyAction>;
  counter: CounterModel;
}

export const createModel = (history: History<any>): RootModel => {
  const rootModel: RootModel = {
    router: reducer(connectRouter(history) as any),
    counter: counterModel,
  };
  return rootModel;
};
