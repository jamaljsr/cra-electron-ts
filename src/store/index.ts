import { routerMiddleware } from 'connected-react-router';
import { createStore, createTypedHooks } from 'easy-peasy';
import { createHashHistory } from 'history';
import { createLogger } from 'redux-logger';
import { createModel, RootModel } from './models';

export const history = createHashHistory();

export const createReduxStore = () => {
  // Redux store Configuration
  const middleware = [];

  // Skip redux logs in console during the tests
  if (process.env.NODE_ENV !== 'test') {
    // Logging Middleware
    const logger = createLogger({
      level: 'info',
      collapsed: true,
    });

    middleware.push(logger);
  }

  // Router Middleware
  const router = routerMiddleware(history);
  middleware.push(router);

  const models = createModel(history);

  // create easy-peasy store
  return createStore(models, {
    middleware,
  });
};

const store = createReduxStore();

// export hooks directly from the store to get proper type inference
const typedHooks = createTypedHooks<RootModel>();
export const { useStoreActions, useStoreDispatch, useStoreState } = typedHooks;

// enable hot reload for models
if (process.env.NODE_ENV === 'development') {
  if ((module as any).hot) {
    (module as any).hot.accept('./models', () => {
      const models = createModel(history);
      store.reconfigure(models);
    });
  }
}

export default store;
