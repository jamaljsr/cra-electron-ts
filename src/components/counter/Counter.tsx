import React, { useEffect } from 'react';
import { Alert, Button, Icon } from 'antd';
import { useStoreState, useStoreActions } from 'store';
import { useAsyncCallback } from 'react-async-hook';
import { useTranslation } from 'react-i18next';
import styles from './Counter.module.less';
import { info } from 'electron-log';

const Counter = () => {
  useEffect(() => info('Rendering Counter component'), []);
  const { t } = useTranslation();
  const { count } = useStoreState(s => s.counter);
  const { increment, decrement, incrementIfOdd, incrementAsync } = useStoreActions(
    a => a.counter,
  );
  const { execute: incrementAsyncCb, loading, error } = useAsyncCallback(() =>
    incrementAsync(),
  );
  const [incrementCb, decrementCb, incrementIfOddCb] = [
    increment,
    decrement,
    incrementIfOdd,
  ].map(x => () => x());

  return (
    <div className={styles.body}>
      {error && <Alert message={error.message} type="error" data-tid="error" />}
      <h1 className={styles.counter} data-tid="counter">
        {loading ? <Icon type="loading" data-tid="async-loader" /> : count}
      </h1>
      <div className={styles.btnGroup}>
        <Button
          type="primary"
          icon="plus"
          className={styles.btn}
          data-tid="incr-btn"
          onClick={incrementCb}
        >
          {t('cmps.counter.increment', 'Increment')}
        </Button>
        <Button
          type="primary"
          icon="minus"
          className={styles.btn}
          data-tid="decr-btn"
          onClick={decrementCb}
        >
          {t('cmps.counter.decrement', 'Decrement')}
        </Button>
        <Button
          type="primary"
          icon="question"
          className={styles.btn}
          data-tid="odd-btn"
          onClick={incrementIfOddCb}
        >
          {t('cmps.counter.increment-odd', 'Increment Odd')}
        </Button>
        <Button
          type="primary"
          icon="retweet"
          className={styles.btn}
          data-tid="async-btn"
          onClick={incrementAsyncCb}
          loading={loading}
        >
          {t('cmps.counter.increment-async', 'Increment Async')}
        </Button>
      </div>
    </div>
  );
};

export default Counter;
