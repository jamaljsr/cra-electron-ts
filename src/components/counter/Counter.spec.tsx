import React from 'react';
import {
  render,
  fireEvent,
  waitForElement,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { StoreProvider } from 'easy-peasy';
import { ConnectedRouter } from 'connected-react-router';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';

import { createReduxStore } from 'store';
import Counter from './Counter';

describe('Counter component', () => {
  const renderComponent = () => {
    const store = createReduxStore();
    const app = (
      <StoreProvider store={store}>
        <Provider store={store as any}>
          <ConnectedRouter history={createMemoryHistory()}>
            <Counter />
          </ConnectedRouter>
        </Provider>
      </StoreProvider>
    );

    return render(app);
  };

  it('should contain default counter value of zero', () => {
    const { getByTestId } = renderComponent();
    expect(getByTestId('counter')).toHaveTextContent('0');
  });

  it('should increment by one when plus btn is clicked', () => {
    const { getByTestId } = renderComponent();
    const btn = getByTestId('incr-btn');
    fireEvent.click(btn);
    expect(getByTestId('counter')).toHaveTextContent('1');
  });

  it('should decrement by one when minus btn is clicked', () => {
    const { getByTestId } = renderComponent();
    const btn = getByTestId('decr-btn');
    fireEvent.click(btn);
    expect(getByTestId('counter')).toHaveTextContent('-1');
  });

  it('should not increment when odd btn is clicked and count is even', () => {
    const { getByTestId } = renderComponent();
    const btn = getByTestId('odd-btn');
    fireEvent.click(btn);
    expect(getByTestId('counter')).toHaveTextContent('0');
  });

  it('should increment by two when odd btn is clicked and count is odd', () => {
    const { getByTestId } = renderComponent();

    // first increment to 1 so that the current count is odd
    const incr = getByTestId('incr-btn');
    fireEvent.click(incr);

    const btn = getByTestId('odd-btn');
    fireEvent.click(btn);
    expect(getByTestId('counter')).toHaveTextContent('3');
  });

  it('should show loader when async btn is clicked', async () => {
    const { getByTestId } = renderComponent();
    const btn = getByTestId('async-btn');
    fireEvent.click(btn);
    const loader = await waitForElement(() => getByTestId('async-loader'));
    expect(loader).not.toBeNull();
  });

  it('should increment after some time when async btn is clicked', async () => {
    const { getByTestId } = renderComponent();
    const btn = getByTestId('async-btn');
    fireEvent.click(btn);

    // wait for loader to show and then be removed
    await waitForElement(() => getByTestId('async-loader'));
    await waitForElementToBeRemoved(() => getByTestId('async-loader'));

    expect(btn).toHaveTextContent('cmps.counter.increment-async');
    expect(getByTestId('counter')).toHaveTextContent('1');
  });

  it('should raise error when count is 3 and async btn is clicked', async () => {
    const { getByTestId } = renderComponent();

    // first increment to 3
    const incr = getByTestId('incr-btn');
    fireEvent.click(incr);
    fireEvent.click(incr);
    fireEvent.click(incr);

    const btn = getByTestId('async-btn');
    fireEvent.click(btn);

    // wait for loader to show and then be removed
    await waitForElement(() => getByTestId('async-loader'));
    await waitForElementToBeRemoved(() => getByTestId('async-loader'));

    expect(btn).toHaveTextContent('cmps.counter.increment-async');
    expect(getByTestId('error')).toHaveTextContent(
      'models.counter.increment-async.error',
    );
  });
});
