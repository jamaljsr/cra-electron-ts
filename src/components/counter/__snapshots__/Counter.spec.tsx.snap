// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component should render the default state of the ui correctly 1`] = `
<div>
  <h1
    data-tid="counter"
  >
    0
  </h1>
  <div>
    <button
      class="ant-btn ant-btn-primary"
      data-tid="incr-btn"
      type="button"
    >
      <i
        aria-label="icon: plus"
        class="anticon anticon-plus"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </i>
      <span>
        Increment
      </span>
    </button>
    <button
      class="ant-btn ant-btn-primary"
      data-tid="decr-btn"
      type="button"
    >
      <i
        aria-label="icon: minus"
        class="anticon anticon-minus"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="minus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </i>
      <span>
        Decrement
      </span>
    </button>
    <button
      class="ant-btn ant-btn-primary"
      data-tid="odd-btn"
      type="button"
    >
      <i
        aria-label="icon: question"
        class="anticon anticon-question"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="question"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M764 280.9c-14-30.6-33.9-58.1-59.3-81.6C653.1 151.4 584.6 125 512 125s-141.1 26.4-192.7 74.2c-25.4 23.6-45.3 51-59.3 81.7-14.6 32-22 65.9-22 100.9v27c0 6.2 5 11.2 11.2 11.2h54c6.2 0 11.2-5 11.2-11.2v-27c0-99.5 88.6-180.4 197.6-180.4s197.6 80.9 197.6 180.4c0 40.8-14.5 79.2-42 111.2-27.2 31.7-65.6 54.4-108.1 64-24.3 5.5-46.2 19.2-61.7 38.8a110.85 110.85 0 0 0-23.9 68.6v31.4c0 6.2 5 11.2 11.2 11.2h54c6.2 0 11.2-5 11.2-11.2v-31.4c0-15.7 10.9-29.5 26-32.9 58.4-13.2 111.4-44.7 149.3-88.7 19.1-22.3 34-47.1 44.3-74 10.7-27.9 16.1-57.2 16.1-87 0-35-7.4-69-22-100.9zM512 787c-30.9 0-56 25.1-56 56s25.1 56 56 56 56-25.1 56-56-25.1-56-56-56z"
          />
        </svg>
      </i>
      <span>
        Increment Odd
      </span>
    </button>
    <button
      class="ant-btn ant-btn-primary"
      data-tid="async-btn"
      type="button"
    >
      <i
        aria-label="icon: retweet"
        class="anticon anticon-retweet"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="retweet"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M136 552h63.6c4.4 0 8-3.6 8-8V288.7h528.6v72.6c0 1.9.6 3.7 1.8 5.2a8.3 8.3 0 0 0 11.7 1.4L893 255.4c4.3-5 3.6-10.3 0-13.2L749.7 129.8a8.22 8.22 0 0 0-5.2-1.8c-4.6 0-8.4 3.8-8.4 8.4V209H199.7c-39.5 0-71.7 32.2-71.7 71.8V544c0 4.4 3.6 8 8 8zm752-80h-63.6c-4.4 0-8 3.6-8 8v255.3H287.8v-72.6c0-1.9-.6-3.7-1.8-5.2a8.3 8.3 0 0 0-11.7-1.4L131 768.6c-4.3 5-3.6 10.3 0 13.2l143.3 112.4c1.5 1.2 3.3 1.8 5.2 1.8 4.6 0 8.4-3.8 8.4-8.4V815h536.6c39.5 0 71.7-32.2 71.7-71.8V480c-.2-4.4-3.8-8-8.2-8z"
          />
        </svg>
      </i>
      <span>
        Increment Async
      </span>
    </button>
  </div>
</div>
`;
